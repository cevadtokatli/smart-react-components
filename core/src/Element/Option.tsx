import styled from 'styled-components'
import applyStyledProps, { StyledProps, shouldForwardProp } from '../styled-props'

export default styled.option.withConfig({ shouldForwardProp })<StyledProps>(applyStyledProps)
