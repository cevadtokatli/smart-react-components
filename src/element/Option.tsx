import styled from "styled-components"
import {Props, attrs, css} from "./"

export default styled.option.attrs<Props>(attrs)<Props>(css)