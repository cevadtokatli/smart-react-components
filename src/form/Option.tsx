import React from "react"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChild, Value, ValueMix, SetValue} from "../props"
import DV from "../default-value"
import OptionElement from "./OptionElement"

export interface Props extends SizeProps {
    elementProps?: JSXElementProps
    children: JSXChild
    value: Value
    searchValue?: string
    disabled?: boolean
    box?: boolean
    type?: string
    fill?: boolean
    shape?: string
    hover?: boolean
    space?: boolean
    active?: ValueMix
    setActive?: SetValue<ValueMix>
}

const Option: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,elementProps=DV.JSX_ELEMENT_PROPS,children,value,disabled,box,type="primary",fill,shape="default",hover,active,setActive}) => {
    /**
     * Returns active$ status of the element. 
     */
    const getActive$ = (): boolean => {
        if(value == null)
            return false
        return Array.isArray(active) ? active.indexOf(value) > -1 : value == active
    }

    const [active$, setActive$] = React.useState<boolean>(() => getActive$())
    
    React.useEffect(() => setActive$(getActive$()), [value, active])

    /**
     * Sets value if element is not disabled.
     */
    const set = (): void => {
        if(disabled || (value == null && Array.isArray(active)))
            return

        if(Array.isArray(active)) {
            if(active.indexOf(value) == -1)
                setActive([...active,value])
            else
                setActive(active.filter(i => i != value))
        } else if(value != active)
            setActive(value)
    }

    return <OptionElement {...elementProps} className={`${active$ ? "src-option-active" : ""}`} size={size} sizeSm={sizeSm} sizeMd={sizeMd} sizeLg={sizeLg} sizeXl={sizeXl} box={box} type={type} fill$={fill} shape={shape} hover={hover} disabled={disabled} value={value} onClick={set}>{children}</OptionElement>
}

export default Option