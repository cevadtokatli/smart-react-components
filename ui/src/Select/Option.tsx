import { ContentElement, PaletteProp, SetState } from '@smart-react-components/core/types'
import React from 'react'
import DropdownItemElement from '../components/Dropdown/DropdownItemElement'
import { FormValue } from '../types'
import WaveEffect from '../WaveEffect'

export interface Props {
  children: ContentElement
  isDisabled?: boolean
  value: FormValue
}

export interface PrivateProps {
  active: FormValue | FormValue[]
  hasHover: boolean
  hasWaveEffect: boolean
  isDisabled: boolean
  isEmbedded: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  setActive: SetState<FormValue | FormValue[]>
  waveEffectPalette: PaletteProp
}

const Option: React.FC<Props> = ({ active, children, hasHover, hasWaveEffect, isDisabled, isOutline, isSoft, palette, setActive, value, waveEffectPalette }: (Props & PrivateProps)) => {
  const handleOnClick = () => {
    if (isDisabled) {
      return
    }

    if (!Array.isArray(active)) {
      setActive(value)
    } else {
      if (!active.includes(value)) {
        setActive([...active, value])
      } else {
        setActive(active.filter(i => i !== value))
      }
    }
  }

  const isActive = () => !Array.isArray(active) ? value === active : active.includes(value)

  const content = (
    <DropdownItemElement
      hasHover={hasHover}
      isActive={isActive()}
      isDisabled={isDisabled}
      isOutline={isOutline}
      isSoft={isSoft}
      palette={palette}
      onClick={handleOnClick}
    >
      {children}
    </DropdownItemElement>
  )

  return hasWaveEffect ? <WaveEffect palette={waveEffectPalette}>{content}</WaveEffect> : content
}

export default Option
