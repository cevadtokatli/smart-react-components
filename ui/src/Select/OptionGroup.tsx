import Div from '@smart-react-components/core/Element/Div'
import { ContentElement } from '@smart-react-components/core/types'
import React from 'react'
import { PrivateProps } from './Option'

export interface Props {
  children: JSX.Element | JSX.Element[]
  label: ContentElement
}

const OptionGroup: React.FC<Props> = ({ children, active, hasHover, hasWaveEffect, isDisabled, isEmbedded, isOutline, isSoft, label, palette, setActive, waveEffectPalette }: Props & PrivateProps) => (
  <>
    <Div fontWeight="$fontWeight.semibold">{label}</Div>
    { (Array.isArray(children) ? children : [children]).map((item, idx) => React.cloneElement(item, {
      key: item.key ?? idx,
      active,
      hasHover,
      hasWaveEffect,
      isEmbedded,
      isOutline,
      isSoft,
      palette,
      setActive,
      waveEffectPalette,
      ...(isDisabled && { isDisabled }),
    })) }
  </>
)

OptionGroup.displayName = 'OptionGroup'

export default OptionGroup
