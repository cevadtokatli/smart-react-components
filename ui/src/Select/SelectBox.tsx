import Div from '@smart-react-components/core/Element/Div'
import { GenericProps } from '../types/form'
import extractElementProps from '@smart-react-components/core/element-props'
import intrinsicStyledCoreProps from '@smart-react-components/core/element-props/intrinsic-styled-core-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { Theme } from '@smart-react-components/core/theme'
import { ContentElement, JSXElementProps } from '@smart-react-components/core/types'
import React from 'react'
import { useTheme } from 'styled-components'
import Badge from '../Badge'
import BadgeIcon from '../Badge/BadgeIcon'
import BadgeList from '../Badge/BadgeList'
import FormBlockLabel from '../components/Form/FormBlockLabel'
import HiddenInput from '../components/Form/HiddenInput'
import InputWrapper from '../components/Input/InputWrapper'
import Dropdown from '../Dropdown'
import DropdownArrowIcon from '../Dropdown/DropdownArrowIcon'
import DropdownListElement from '../components/Dropdown/DropdownListElement'
import useAddons from '../hooks/useAddons'
import useSelectBoxItemList from '../hooks/useSelectBoxItemList'
import CloseIcon from '../icons/Close'
import InputElement from '../components/Input/InputElement'
import InputPlaceholder from '../components/Input/InputPlaceholder'
import useSelectBoxHover from '../hooks/useSelectBoxHover'
import { FormValue } from '../types'
import { getInputValue } from '../util/form'
import { applyResponsiveStyledProp } from '../util/props'
import { getWaveEffectPalette } from '../util/wave-effect'
import InputAddon from './SelectAddon'

export interface Props extends GenericProps {
  dropdownArrowIconElementProps?: JSXElementProps
  placeholder?: string
}

const SelectBox = React.forwardRef<HTMLInputElement, Props>((props, forwardRef) => {
  const theme = useTheme() as Theme

  const formEl = React.useRef<HTMLDivElement>(null)

  const { itemList } = useSelectBoxItemList({ children: props.children })

  const handleBadgeClick = value => props.setActive((props.active as FormValue[]).filter(i => i !== value))

  const content = React.useMemo<ContentElement>(() => {
    if (
      typeof props.active === 'undefined'
      || props.active === null
      || (Array.isArray(props.active) && props.active.length === 0)
      || (props.active === '' && !itemList[''])
    ) {
      return <InputPlaceholder>{props.placeholder}</InputPlaceholder>
    }

    if (!Array.isArray(props.active)) {
      return <div>{itemList[String(props.active)]?.children}</div>
    }

    return (
      <BadgeList
        palette={props.palette}
        isSoft={props.isSoft}
        shape={props.shape}
        size={props.size}
        sizeSm={props.sizeSm}
        sizeMd={props.sizeMd}
        sizeLg={props.sizeLg}
        sizeXl={props.sizeXl}
      >
        {
          props.active
            .map(item => item)
            .sort((a, b) => itemList[String(a)].idx < itemList[String(b)].idx ? -1 : 1)
            .map(item => (
              <Badge key={String(item)} elementProps={{ 'data-src-not-clickable': true }}>
                <BadgeIcon onClick={() => handleBadgeClick(item)}>
                  <CloseIcon />
                </BadgeIcon>
                {itemList[String(item)].children}
              </Badge>
            ))
        }
      </BadgeList>
    )
  }, [props.active, props.placeholder, itemList])

  const [dropdownStatus, setDropdownStatus] = React.useState(false)

  const waveEffectPalette = React.useMemo(() => getWaveEffectPalette(props, theme.$.vars.isDarkMode), [props.waveEffectPalette, props.palette, props.isOutline, props.isSoft, theme.$.vars.isDarkMode])

  const { hovered, setHovered } = useSelectBoxHover({
    active: props.active,
    children: Array.isArray(props.children) ? props.children : [props.children],
    dropdownStatus,
    hasHover: props.hasHover,
    isDisabled: props.isDisabled,
    setActive: props.setActive,
    setDropdownStatus,
  })

  useChangeEffect(() => {
    const event = new Event('src.fixedBox.setPosition')
    formEl.current.dispatchEvent(event)
  }, [content])

  const { leftAddon, rightAddon } = useAddons({
    Component: InputAddon,
    leftAddon: props.leftAddon,
    rightAddon: props.rightAddon,
    props: {
      cursorKey: 'selectBox',
      hasBorder: props.hasBorder,
      isDisabled: props.isDisabled,
      isFocused: false,
      isOutline: props.isOutline,
      isSoft: props.isSoft,
      palette: props.palette,
      shape: props.shape,
      size: props.size,
      sizeSm: props.sizeSm,
      sizeMd: props.sizeMd,
      sizeLg: props.sizeLg,
      sizeXl: props.sizeXl,
    },
  })

  const handleOptionClick = active => {
    props.setActive(active)

    if (!Array.isArray(active)) {
      setDropdownStatus(false)
    }
  }

  return (
    <Dropdown
      elementProps={props.elementProps}
      status={dropdownStatus}
      setStatus={setDropdownStatus}
    >
      <FormBlockLabel
        {...extractElementProps(props, [intrinsicStyledCoreProps])}
        as="div"
        cursorKey="selectBox"
        formSize={props.size}
        formSizeSm={props.sizeSm}
        formSizeMd={props.sizeMd}
        formSizeLg={props.sizeLg}
        formSizeXl={props.sizeXl}
        isBlock={props.isBlock}
        isDisabled={props.isDisabled}
        ref={formEl}
      >
        { props.label && <span>{props.label}</span> }
        <Div
          display="flex"
          flex="1 1 auto"
        >
          { leftAddon?.props?.isSeparated && leftAddon }
          <InputWrapper
            cursor="$cursor.selectBox"
            hasBorder={props.hasBorder}
            hasSeparatedLeftAddon={leftAddon?.props?.isSeparated}
            hasSeparatedRightAddon={rightAddon?.props?.isSeparated}
            isDisabled={props.isDisabled}
            isFocused={false}
            isOutline={props.isOutline}
            isSoft={props.isSoft}
            palette={props.palette}
            shape={props.shape}
          >
            { (leftAddon && !leftAddon?.props?.isSeparated) && leftAddon }
            <InputElement
              hasLeftAddon={!!leftAddon}
              hasRightAddon={!!rightAddon}
              inputSize={props.size}
              inputSizeSm={props.sizeSm}
              inputSizeMd={props.sizeMd}
              inputSizeLg={props.sizeLg}
              inputSizeXl={props.sizeXl}
              isDisabled={props.isDisabled}
              isFocused={false}
              isInput={false}
              isOutline={props.isOutline}
              isSoft={props.isSoft}
              palette={props.palette}
            >
              <HiddenInput
                onChange={() => {}}
                ref={forwardRef}
                value={getInputValue(props.active)}
                {...(props.isDisabled && { disabled: true })}
                {...(props.isRequired && { required: true })}
              />
              { content }
              <DropdownArrowIcon
                fill="currentcolor"
                {...applyResponsiveStyledProp(props, 'size', 'iconSize', v => `$size.icon.${v}`)}
                {...props.dropdownArrowIconElementProps}
              />
            </InputElement>
            { (rightAddon && !rightAddon?.props?.isSeparated) && rightAddon }
          </InputWrapper>
          { rightAddon?.props?.isSeparated && rightAddon }
        </Div>
      </FormBlockLabel>
      <DropdownListElement
        dropdownListSize={props.size}
        dropdownListSizeSm={props.sizeSm}
        dropdownListSizeMd={props.sizeMd}
        dropdownListSizeLg={props.sizeLg}
        dropdownListSizeXl={props.sizeXl}
        isOutline={props.isOutline}
        isSoft={props.isSoft}
        palette={props.palette}
      >
        { (Array.isArray(props.children) ? props.children : [props.children]).map((item, idx) => item && React.cloneElement(item, {
          key: item.key ?? idx,
          active: props.active,
          cursorKey: 'selectBox',
          hasHover: false,
          hasWaveEffect: props.hasWaveEffect,
          hovered,
          isEmbedded: false,
          isOutline: props.isOutline,
          isSoft: props.isSoft,
          palette: props.palette,
          setActive: handleOptionClick,
          setHovered: props.hasHover ? setHovered : null,
          waveEffectPalette,
          ...(props.isDisabled && { isDisabled: true }),
        })) }
      </DropdownListElement>
    </Dropdown>
  )
})

SelectBox.defaultProps = {
  dropdownArrowIconElementProps: {},
  elementProps: {},
  hasBorder: true,
  hasHover: true,
  hasWaveEffect: true,
  isBlock: true,
  isOutline: true,
  palette: 'primary',
  shape: 'rectangle',
  size: 'medium',
}

export default SelectBox
