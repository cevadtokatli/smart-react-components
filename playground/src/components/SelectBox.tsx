import SelectBox from "@smart-react-components/ui/Select/SelectBox"
import Option from "@smart-react-components/ui/Select/Option"
import OptionGroup from "@smart-react-components/ui/Select/OptionGroup"
import React from 'react'

export const Palette = () => {
  const [active, setActive] = React.useState(() => null)

  return (
    <>
      <SelectBox active={active} setActive={setActive} placeholder="Label">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="secondary" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="success" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="danger" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="warning" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="info" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="dynamic" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="!dynamic" active={active} setActive={setActive}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}

export const Soft = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <SelectBox active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="secondary" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="success" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="danger" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="warning" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="info" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="dynamic" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="!dynamic" active={active} setActive={setActive} isSoft>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}

export const Outline = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <SelectBox active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="secondary" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="success" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="danger" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="warning" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="info" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="dynamic" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="!dynamic" active={active} setActive={setActive} isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}

export const SoftOutline = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <SelectBox active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="secondary" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="success" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="danger" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="warning" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="info" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="dynamic" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox palette="!dynamic" active={active} setActive={setActive} isSoft isOutline={false}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}

export const Size = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <SelectBox active={active} setActive={setActive} size="small" label="Small">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox active={active} setActive={setActive} label="Medium">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox active={active} setActive={setActive} size="large" label="Large">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}

export const Shape = () => {
  const [active, setActive] = React.useState(() => [])

  return (
    <>
      <SelectBox active={active} setActive={setActive} label="Rectangle">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
      <SelectBox active={active} setActive={setActive} label="Rounded" shape="rounded">
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    </>
  )
}
