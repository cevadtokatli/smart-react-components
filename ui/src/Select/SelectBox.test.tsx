import { wrapTheme } from '@smart-react-components/core/test'
import { fireEvent, render, waitFor } from '@testing-library/react' 
import React from 'react'
import SelectBox from './SelectBox'
import Option from './Option'
import OptionGroup from './OptionGroup'

describe('<SelectBox />', () => {
  it('should render component', async () => {
    const { asFragment, container } = render(wrapTheme(
      <SelectBox active="item-1" setActive={jest.fn()}>
        <OptionGroup label="Category">
          <Option value="item-1">Item 1</Option>
          <Option value="item-2">Item 2</Option>
          <Option value="item-3">Item 3</Option>
        </OptionGroup>
      </SelectBox>
    ))

    await waitFor(() => {
      fireEvent.click(container.querySelector('div')!)
    })
  
    expect(asFragment()).toMatchSnapshot()
  })
})
