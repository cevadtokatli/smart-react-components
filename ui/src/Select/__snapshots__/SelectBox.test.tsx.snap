// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBox /> should render component 1`] = `
<DocumentFragment>
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c0 > span {
  margin-bottom: .5rem;
}

.c1 > span {
  display: block;
  word-break: break-word;
  word-wrap: break-word;
  -webkit-tap-highlight-color: transparent;
}

.c5 {
  height: 100%;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  top: 0;
  width: 100%;
}

.c6 {
  fill: currentcolor;
  height: 1rem;
  width: 1rem;
}

.c3 {
  font-size: 1rem;
  padding: .45rem .9rem;
}

.c4 {
  border: solid 1px;
  border-radius: .2rem;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  font-family: inherit;
  -webkit-transition: 200ms 0s ease-in-out;
  transition: 200ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  -webkit-appearance: none;
  -webkit-tap-highlight-color: transparent;
  background: #fff;
  border-color: #dbdbdb;
  color: #323a46;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
}

.c4:active,
.c4:focus {
  outline: 0;
}

@media (max-width:767px) {

}

@media (max-width:767px) {

}

@media (max-width:767px) {

}

@media (min-width:768px) {

}

<div
    class="c0 c1"
  >
    <div
      class="c2"
      display="flex"
    >
      <div
        class="c3 c4"
        shape="rectangle"
      >
        <input
          class="c5"
          value="item-1"
        />
        <div>
          Item 1
        </div>
        <svg
          class="c6"
          fill="currentcolor"
          height="$size.icon.medium"
          viewBox="5 4 14 14"
          width="$size.icon.medium"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
        </svg>
      </div>
    </div>
  </div>
  .c2 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
}

.c3 {
  overflow: initial !important;
}

.c5 {
  padding: .3125rem 0;
}

.c5 > div {
  font-size: .875rem;
  margin: .125rem .45rem;
  padding: .3125rem .3125rem;
}

.c7 {
  font-weight: 500;
}

.c8 {
  overflow: hidden;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c0 {
  position: fixed;
  z-index: 2000;
}

.c4 {
  overflow: auto;
  position: fixed;
  z-index: 1000;
}

.c6 {
  border-radius: .2rem;
  box-shadow: 0 0 10px 1px rgba(154,161,171,.15);
  font-family: inherit;
  height: inherit;
  overflow: auto;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: inherit;
  -webkit-overflow-scrolling: touch;
  background: #fff;
  color: #323a46;
  fill: #323a46;
}

.c6 > div {
  word-break: break-word;
  word-wrap: break-word;
}

.c9 {
  border-radius: .2rem;
  cursor: default;
  -webkit-transition: 100ms 0s ease-in-out;
  transition: 100ms 0s ease-in-out;
  -webkit-transition-property: background,color,fill;
  transition-property: background,color,fill;
  background: #0553AC;
  color: #fff;
}

.c10 {
  border-radius: .2rem;
  cursor: default;
  -webkit-transition: 100ms 0s ease-in-out;
  transition: 100ms 0s ease-in-out;
  -webkit-transition-property: background,color,fill;
  transition-property: background,color,fill;
}

.c10:hover {
  background: #0665d0;
  color: #fff;
}

@media (max-width:767px) {
  .c0 {
    box-sizing: border-box;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    height: 100%;
    left: 0;
    padding: 1.75rem;
    top: 0;
    touch-action: none;
    width: 100%;
    background: rgba(0,0,0,.3);
  }
}

@media (max-width:767px) {
  .c4 {
    height: auto !important;
    max-height: 100% !important;
    position: static;
    width: 100% !important;
  }
}

@media (max-width:767px) {
  .c1 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    will-change: opacity;
  }

  .c1.src-fixed-box-show {
    opacity: 0;
  }

  .c1.src-fixed-box-show-active {
    opacity: 1;
  }

  .c1.src-fixed-box-hide {
    opacity: 1;
  }

  .c1.src-fixed-box-hide-active {
    opacity: 0;
  }

  .c1.src-fixed-box-show-active,
  .c1.src-fixed-box-hide-active {
    -webkit-transition: opacity 300ms 0s ease-in-out;
    transition: opacity 300ms 0s ease-in-out;
  }
}

@media (min-width:768px) {
  .c1 > * {
    will-change: opacity,transform;
  }

  .c1.src-fixed-box-show > * {
    opacity: 0;
    -webkit-transform: translateY(60px);
    -ms-transform: translateY(60px);
    transform: translateY(60px);
  }

  .c1.src-fixed-box-show-active > * {
    opacity: 1;
    -webkit-transform: translateY(0);
    -ms-transform: translateY(0);
    transform: translateY(0);
  }

  .c1.src-fixed-box-hide > * {
    opacity: 1;
    -webkit-transform: translateY(0);
    -ms-transform: translateY(0);
    transform: translateY(0);
  }

  .c1.src-fixed-box-hide-active > * {
    opacity: 0;
    -webkit-transform: translateY(60px);
    -ms-transform: translateY(60px);
    transform: translateY(60px);
  }

  .c1.src-fixed-box-show-active > *,
  .c1.src-fixed-box-hide-active > * {
    -webkit-transition: 300ms 0s ease-in-out;
    transition: 300ms 0s ease-in-out;
    -webkit-transition-property: opacity,-webkit-transform;
    -webkit-transition-property: opacity,transform;
    transition-property: opacity,transform;
  }
}

<div
    class=" c0 c1 sc-bczRLJ sc-ftvSup  c0"
  >
    <div
      class="c2"
      height="100%"
      width="100%"
    />
    <div
      class="c3 c4"
      overflow="initial,true"
      style="pointer-events: none; visibility: hidden;"
    >
      <div
        class="c5 c6"
      >
        <div
          class="c7"
          font-weight="$fontWeight.semibold"
        >
          Category
        </div>
        <div
          class="c8 c9"
          overflow="hidden"
        >
          Item 1
        </div>
        <div
          class="c8 c10"
          overflow="hidden"
        >
          Item 2
        </div>
        <div
          class="c8 c10"
          overflow="hidden"
        >
          Item 3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
